### ISSUE TITLE
Specifying encryption to use a subkey rather than the primary.

### ISSUE THREAD

**Issue by nigelainscoe** [ProtonMail/gopenpgp#159]
I am trying to encrypt using a client's public key where the primary key is DSA which I understand is only for signing. It also has an ELG-E subkey. I have spent the morning searching but to no avail. Is there something I have missed in order to be able to get the code to use the subkey rather than the primary key? 
 
Please complete the following checklist (by adding [x]):
- [x ] I have searched open and closed issues for duplicates
- [x ] This isn't a feature request
- [x ] This is not a report about my app not working as expected

----------------------------------------

**Comment by twiss** [comment]
Gopenpgp should automatically use the encryption subkey when encrypting. Are you sure it doesn't? If so, that's a bug, could you post an example key and resulting message?

**Comment by nigelainscoe** [comment]
Unfortunately, the key in question is a client key so even though it is a public key, I'm not prepared to post it here. I have had a look online and it doesn't appear to be in the public domain.

Fun fact, I have attempted to reproduce the error with a key of my own but have not been able to do so. 
The only difference between my key and the clients is that the client key has a DSA key strength of 1024, whereas my version of Kleopatra will only let me create it with a key strength of 2048. I'm going to see if I can force gpg to make me one from the command line though.  If I can and it reproduces the error, I'll post the key.

Their key:
![image](https://user-images.githubusercontent.com/822636/147576875-b8e683f6-bca3-4035-9f1a-c16682c902fa.png)

My key:
![image](https://user-images.githubusercontent.com/822636/147577029-632462a6-6b34-4fef-9412-0b5fc9505d06.png)

Running the simplest:
```
	armor, err := helper.EncryptMessageArmored(PublicKey, PlainText)
	if err != nil {
		fmt.Println("messed up:")
	}

```
Produces the following in the error object. You can see the type refers to PubKeyAlgoDSA (17)
```
msg = gopenpgp: unable to encrypt message
cause = cannot encrypt to public key of type 17
msg = gopenpgp: error in encrypting asymmetrically

```

And yes I will be pointing out that having a 10-year-old key with no expiry date is not really setting the best example for a regulated financial institution 😉

**Comment by twiss** [comment]
This is a bit of a random guess, but could you check whether the key has the "encrypt communications" flag set, and not just the "encrypt storage" flag?

If it does, could you perhaps post the output of `gpg --list-packets <key file>`, or `sq packet dump <key file>`, perhaps anonymized?

**Comment by nigelainscoe** [comment]
Hope this helps.

```
# off=0 ctb=99 tag=6 hlen=3 plen=418                               
:public key packet:                                                
        version 4, algo 17, created 1301578632, expires 0          
        pkey[0]: [1024 bits]                                       
        pkey[1]: [160 bits]                                        
        pkey[2]: [1024 bits]                                       
        pkey[3]: [1024 bits]                                       
        keyid: 9D8FED1F02D81C18                                    
# off=421 ctb=b4 tag=13 hlen=2 plen=43                             
:user ID packet: "user <user@example.com>"     
# off=466 ctb=88 tag=2 hlen=2 plen=70                              
:signature packet: algo 17, keyid 9D8FED1F02D81C18                 
        version 4, created 1301578632, md5len 0, sigclass 0x10     
        digest algo 8, begin of digest dd 07                       
        hashed subpkt 2 len 4 (sig created 2011-03-31)             
        subpkt 16 len 8 (issuer key ID 9D8FED1F02D81C18)           
        data: [159 bits]                                           
        data: [155 bits]                                           
# off=538 ctb=b9 tag=14 hlen=3 plen=525                            
:public sub key packet:                                            
        version 4, algo 16, created 1301578646, expires 0          
        pkey[0]: [2048 bits]                                       
        pkey[1]: [2 bits]                                          
        pkey[2]: [2047 bits]                                       
        keyid: DBD3D35BB1ABCC42                                    
# off=1066 ctb=88 tag=2 hlen=2 plen=70                             
:signature packet: algo 17, keyid 9D8FED1F02D81C18                 
        version 4, created 1301578646, md5len 0, sigclass 0x18     
        digest algo 8, begin of digest bb 4c                       
        hashed subpkt 2 len 4 (sig created 2011-03-31)             
        subpkt 16 len 8 (issuer key ID 9D8FED1F02D81C18)           
        data: [156 bits]                                           
        data: [160 bits]                                           

```
I'm trying to debug the code but it's slow going.

**Comment by twiss** [comment]
According to this listing, the key does not have any usages specified at all, neither for the primary key nor the subkey. The usages that Kleopatra reports ("Certify, Sign, Authenticate" and "Encrypt", respectively) are perhaps just the defaults it uses when the usages are absent; the mentioned usages are not present in the key. Also, note that the [spec says](https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.3.21):

> If a list is shorter than an implementation expects, the unstated flags are considered to be zero.

In other words, by default we are to assume that the key may not be used for encryption, or anything else, really. So I think gopenpgp is behaving per spec (except that it is, apparently, willing to try encrypting to the primary key even without any flags, which then of course fails, leading to the weird error message you're seeing).

**Comment by nigelainscoe** [comment]
Yea, I've just this second tracked it down to here and subkey.Sig.FlagEncryptCommunications is indeed false - as is the rest.
```
	// Iterate the keys to find the newest, unexpired one
	candidateSubkey := -1
	var maxTime time.Time
	for i, subkey := range e.Subkeys {
		if subkey.Sig.FlagsValid &&
			subkey.Sig.FlagEncryptCommunications &&
			subkey.PublicKey.PubKeyAlgo.CanEncrypt() &&
			!subkey.PublicKey.KeyExpired(subkey.Sig, now) &&
			!subkey.Sig.SigExpired(now) &&
			!subkey.Revoked(now) &&
			(maxTime.IsZero() || subkey.Sig.CreationTime.After(maxTime)) {
			candidateSubkey = i
			maxTime = subkey.Sig.CreationTime
		}
	}

```

**Comment by nigelainscoe** [comment]
Thanks, yes a better message might have helped - although "The public key you are trying to use is not valid" might have confused me even more given I can encrypt stuff with Kleopatra using that key and the client can happily decrypt the output!

Thank you for your help today, it was most welcome. Feel free to close the issue.

**Comment by nigelainscoe** [comment]
Also, while I've got your attention, I was looking for where one might specify multiple keys for an encryption task, i.e. encrypt with their key and my key so I can decrypt in case of dispute? I've looked through the docs and done much Googling but find nothing of any use.
`write.go.Encrypt(ciphertext io.Writer, to []*Entity,......`  looks like it might be promising since it seems to accept an array of entities.

**Comment by twiss** [comment]
You can encrypt to a keyring: https://pkg.go.dev/github.com/ProtonMail/gopenpgp/v2@v2.4.0/crypto#NewKeyRing
